File: python/2022/04/main.py
#!/usr/bin/env python3

import sys
sys.path.append("../../")
sys.path.append("../")
import util as u

###############
debug = False
###############

file = "sample.txt" if debug else "input.txt"

def read_input():
    return u.read_match(file,"(.*)-(.*),(.*)-(.*)")

def part1():
    A = read_input()
    k = sum([1 if (int(a)<=int(c) and int(b)>=int(d))or (int(a)>=int(c) and int(b)<=int(d)) else 0 for a,b,c,d in A])
    print("Part 1:",k)

def part2():
    A = read_input()
    k = sum([1 if not ((int(b)<int(c) )or (int(a)>int(d))) else 0 for a,b,c,d in A])
    print("Part 2:",k)

part1()
part2()

File: c/2022/04/main.c
#include "../../util.h"

#if 0
static char *FILE_NAME = "sample.txt";
#else
static char *FILE_NAME = "input.txt";
#endif


void part1() {
    ARR_MUL_INTEGERS* in = readIntInputDelim(FILE_NAME,"-,");
    printf("%d\n",in->length);
    int64_t k = 0;
    for (int64_t i = 0; i < in->length; i++) {
        if ((in->integers[i].integers[0] <= in->integers[i].integers[2] && in->integers[i].integers[1] >= in->integers[i].integers[3]) ||
            (in->integers[i].integers[0] >= in->integers[i].integers[2] && in->integers[i].integers[1] <= in->integers[i].integers[3]))
            k++;
    }
    printf("Part 1: %ld\n", k);
    freeArrMulInteger(in);
}

void part2() {
    ARR_MUL_INTEGERS* in = readIntInputDelim(FILE_NAME,"-,");
    int64_t k = 0;    
    for (int64_t i = 0; i < in->length; i++) {
        if (in->integers[i].integers[1] >= in->integers[i].integers[2] && in->integers[i].integers[0] <= in->integers[i].integers[3])
            k++;
    }
    printf("Part 2: %ld\n", k);
    freeArrMulInteger(in);
}

int main(void) {
    part1();
    part2();
    return 0;
}
